### PR TITLE
rustdoc: reference function signature types from the `p` array

### PR DESCRIPTION
**Pull request by notriddle**
This reduces the size of the function signature index, because it's common to have many functions that operate on the same types.

    $ wc -c search-index-old.js search-index-new.js
    5224374 search-index-old.js
    3932314 search-index-new.js

By my math, this reduces the uncompressed size of the search index by 32%.
On compressed signatures, the wins are less drastic, a mere 8%:

    $ wc -c search-index-old.js.gz search-index-new.js.gz
    404532 search-index-old.js.gz
    371635 search-index-new.js.gz

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Changes look good to me. Do you have an online version by any chance?

File: src/librustdoc/html/static/js/search.js [hunk 1]
```diff
@@ -1879,7 +1935,7 @@ function initSearch(rawSearchIndex) {
              *   d: Array<string>,
              *   q: Array<string>,
              *   i: Array<Number>,
-             *   f: Array<Array<?>>,
+             *   f: Array<0 | Object>,
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1938]: Could you declare the type returned from `buildFunctionSearchType` in `externs.js` so you can show it here as well please?

File: src/librustdoc/html/static/js/search.js [hunk 2]
```diff
@@ -1836,6 +1825,65 @@ function initSearch(rawSearchIndex) {
             filterCrates);
     }
 
+    function buildItemSearchTypeAll(types, lowercasePaths) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1828]: Please add documentation here and on `buildFunctionSearchType`.

File: src/librustdoc/html/static/js/search.js [hunk 3]
```diff
@@ -984,23 +977,22 @@ function initSearch(rawSearchIndex) {
           *                     no match, returns `MAX_LEV_DISTANCE + 1`.
           */
         function checkType(row, elem, literalSearch) {
-            if (row[NAME].length === 0) {
+            if (row.name === null) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:980]: Maybe I missed it but I only saw you setting `name` to `null` manually and never checking its length. Shouldn't it insert `null` in both "build functions" if the name is empty?

**Comment by Folyd** [comment]
This is so good. I hope this PR wouldn't break my extension, haha. 😄

File: src/librustdoc/html/static/js/externs.js [hunk 4]
```diff
@@ -81,3 +81,56 @@ let ResultsTable;
  * }}
  */
 let Results;
+
+/**
+ * A pair of [inputs, outputs], or 0 for null. This is gets stored in the search index.
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/externs.js:86]: Little typo:

```suggestion
 * A pair of [inputs, outputs], or 0 for null. This is stored in the search index.
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/externs.js:86]: Btw, why `0` and not `null`?

**Comment by GuillaumeGomez** [comment]
Still this question:

> Btw, why `0` and not `null`?

Apart from that, looks really great. :)

**Comment by GuillaumeGomez** [comment]
Please add this explanation then. :)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks, it's awesome! r=me once CI pass.

**Pull request merged**